test/extended/router/external_certificate.go

```diff
@@ -0,0 +1,769 @@
+				o.Expect(err).NotTo(o.HaveOccurred())
+				_, err = oc.KubeClient().CoreV1().Secrets(oc.Namespace()).Create(context.Background(), secret, metav1.CreateOptions{})
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				g.By("Creating a route")
+				route := generateRouteWithExternalCertificate(oc.Namespace(), "route", secret.Name, helloPodSvc, host, routev1.TLSTerminationEdge)
+				// Add inline certificate
+				route.Spec.TLS.Certificate = "my-crt"
+				_, err = oc.RouteClient().RouteV1().Routes(oc.Namespace()).Create(context.Background(), route, metav1.CreateOptions{})
+				o.Expect(err).To(o.HaveOccurred())
+				o.Expect(err.Error()).To(o.ContainSubstring(`Invalid value: "my-tls-secret": cannot specify both tls.certificate and tls.externalCertificate`))
+			})
+		})
+	})
+
+	g.Context("with valid setup", func() {
+		var (
+			secret       *corev1.Secret
+			routes       []*routev1.Route
+			hosts        []string
+			rootDerBytes []byte
+		)
+
+		g.BeforeEach(func() {
+			const numRoutes = 5
+			var routeNames []string
+
+			for i := 0; i < numRoutes; i++ {
+				hosts = append(hosts, fmt.Sprintf("host-%d-%s.%s", i, oc.Namespace(), defaultDomain))
+				routeNames = append(routeNames, fmt.Sprintf("route-%d", i))
+			}
+
+			g.By("Creating a TLS certificate secret")
+			secret, rootDerBytes, err = generateTLSCertSecret(oc.Namespace(), "my-tls-secret", corev1.SecretTypeTLS, hosts...)
+			o.Expect(err).NotTo(o.HaveOccurred())
+			_, err = oc.KubeClient().CoreV1().Secrets(oc.Namespace()).Create(context.Background(), secret, metav1.CreateOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Providing router service account permissions to get,list,watch the secret")
+			_, err = oc.KubeClient().RbacV1().Roles(oc.Namespace()).Create(context.Background(),
+				generateSecretReaderRole(oc.Namespace(), "my-tls-secret"), metav1.CreateOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+			_, err = oc.KubeClient().RbacV1().RoleBindings(oc.Namespace()).Create(context.Background(),
+				generateRouterRoleBinding(oc.Namespace()), metav1.CreateOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By("Creating multiple routes referencing same external certificate")
+			for i := 0; i < numRoutes; i++ {
+				route := generateRouteWithExternalCertificate(oc.Namespace(), routeNames[i], secret.Name, helloPodSvc, hosts[i], routev1.TLSTerminationEdge)
+				_, err = oc.RouteClient().RouteV1().Routes(oc.Namespace()).Create(context.Background(), route, metav1.CreateOptions{})
+				o.Expect(err).NotTo(o.HaveOccurred())
+				routes = append(routes, route)
+			}
+		})
+
+		g.Describe("the router should support external certificate", func() {
+			g.It("and routes are reachable", func() {
+				g.By("Sending https request")
+				for _, route := range routes {
+					hostName, err := getHostnameForRoute(oc, route.Name)
+					o.Expect(err).NotTo(o.HaveOccurred())
+					resp, err := httpsGetCall(hostName, rootDerBytes)
```

Comment:
We can generate the `caPool` once (from `rootDerBytes`) and pass it to `httpsGetCall()` instead of doing it for each route.

Upd: I think I see why it was done this way, the "update secret" test would change the certificate, so we will have to regenerate caPool later.

Reply:
Btw would 2 routes be sufficient to test the shared secret case?

Reply:
> Btw would 2 routes be sufficient to test the shared secret case?

We actually have 5 routes to test the shared secret case. `const numRoutes = 5`. 

xRef: https://github.com/openshift/origin/pull/29499/files#diff-ab54e60f94a61345960f23ba755e95aadcb3a00287fde5d154c08c65009abf0dR174

Let me increase it to `6` since [contention tracker](https://github.com/openshift/router/blob/b41f9d05467fb7b3f6c2dafa6ac4b5e25164c0b6/pkg/router/controller/contention.go#L86) is also at `5`

Reply:
As discussed, let's add a comment explaining the number of the routes.

Reply:
Sure, added a detailed comment explaining the reasoning behind it.